### ISSUE TITLE
Calling .forRoot in a shared module is BAD BAD BAD

### ISSUE THREAD

**Issue by exedor** [akserg/ng2-dnd#108]
* **I'm submitting a ...**
[ X ] bug report
[ ] feature request
[ ] question about the decisions made in the repository

* **Do you want to request a *feature* or report a *bug*?**
Report a bug

* **What is the current behavior?**
You are suggesting that users call .forRoot() in a shared module when using shared modules. This is contrary the way angular is intended to work. If you are instantiating services this can lead to memory leaks, depending on your implementation. Even if safe with your implementation, you should be providing a .forChild() method in your module and advising users to call that, even if for now, it simply makes an internal call back to .forRoot(). This is especially an issue for lazy loaded feature modules. 

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** via https://plnkr.co or similar.
N/A

* **What is the expected behavior?**
.forRoot() is called ONCE each time the app loads, typically in a root module.
.forChild() is called repeatedly in both eager and lazy loaded modules. The lazy loaded modules can get their own instances of services (required for lazy loading) and eager loaded modules can be happy using the app wide scoped services.

* **What is the motivation / use case for changing the behavior?**
Good design, quality software, and compatibility with angular2 applications.

* **Please tell us about your environment:**

- Angular version: 2.3.1
- Browser: all

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, gitter, etc)
N/A

**Comment by arat3091** [comment]
Moreover i think if we put .forRoot in shared module then as this shared module is imported in other modules , we will get new services instances in each of the imported module. We should just need one instance of services in our application.

**Comment by exedor** [comment]
Correct, as I mentioned, that's where the memory leak comes into play. However, lazy loaded modules are the notable exception. They get their own injector instantiated at module load time. This is why you need a .forChild so that problem can be managed and the appropriate instance returned depending on the circumstance.

I've had to deal with this using ng2-bootstrap, ng2-translate, and now this one. I'm finding this is a common pitfall among module developers.

**Comment by arat3091** [comment]
i think the way i understand is that
forRoot creates a module that contains all the directives, the given routes, and the service itself.

forChild creates a module that contains all the directives and the given routes, but does not include the service.

So irrespective of lazy module or eager module registering service twice can cause run time problems.

**Comment by exedor** [comment]
I'd take a look at the angular router documentation to be sure that is correct. .forRoot does instantiate everything including the service(s). Technically, since lazy loaded modules have their own instance of the angular2 injector, it *might* be required to actually re-instantiate another service, but only in the lazy loaded module case.

Alternatively, it may be the case, that forChild need only give the new injector a handle the existing instance of the service created when calling .forRoot.

I wish I knew enough about it at this point to tell you exactly how it is supposed to work. I only know the way it is working, again depending on your underlying implementation, is a problem...especially if services are involved. The angular2 docs make an extra effort to specifically call out the fact that app devs should only be calling .forRoot in their root module and nowhere else. They also make an extra point of mentioning that lazy loaded modules get their own injector instance. They leave it as an exercise to the reader to connect the dots and realize that means they are going to have problems with lazy loaded modules and libraries like this that don't handle service loading correctly.

The angular2 router for instance, has you call .forRoot with master routes in the root module. Then in feature modules (again, regardless of lazy or eager loading) has you call .forChild with the child routes that you're loading in feature/child modules.

The angular2 router is almost certainly using services and that module properly handles this situation.
